core/generator.js

```diff
@@ -29,390 +29,508 @@ const {Workspace} = goog.requireType('Blockly.Workspace');
+        // Value blocks return tuples of code and operator order.
+        // Top-level blocks don't care about operator order.
+        line = line[0];
+      }
+      if (line) {
+        if (block.outputConnection) {
+          // This block is a naked value.  Ask the language's code generator if
+          // it wants to append a semicolon, or something.
+          line = this.scrubNakedValue(line);
+          if (this.STATEMENT_PREFIX && !block.suppressPrefixSuffix) {
+            line = this.injectId(this.STATEMENT_PREFIX, block) + line;
+          }
+          if (this.STATEMENT_SUFFIX && !block.suppressPrefixSuffix) {
+            line = line + this.injectId(this.STATEMENT_SUFFIX, block);
+          }
+        }
+        code.push(line);
+      }
+    }
+    code = code.join('\n');  // Blank line between each section.
+    code = this.finish(code);
+    // Final scrubbing of whitespace.
+    code = code.replace(/^\s+\n/, '');
+    code = code.replace(/\n\s+$/, '\n');
+    code = code.replace(/[ \t]+\n/g, '\n');
+    return code;
+  }
+
+  // The following are some helpful functions which can be used by multiple
+
+  // languages.
 
   /**
-   * Maximum length for a comment before wrapping.  Does not account for
-   * indenting level.
-   * @type {number}
+   * Prepend a common prefix onto each line of code.
+   * Intended for indenting code or adding comment markers.
+   * @param {string} text The lines of code.
+   * @param {string} prefix The common prefix.
+   * @return {string} The prefixed lines of code.
    */
-  this.COMMENT_WRAP = 60;
+  prefixLines(text, prefix) {
+    return prefix + text.replace(/(?!\n$)\n/g, '\n' + prefix);
+  }
 
   /**
-   * List of outer-inner pairings that do NOT require parentheses.
-   * @type {!Array<!Array<number>>}
+   * Recursively spider a tree of blocks, returning all their comments.
+   * @param {!Block} block The block from which to start spidering.
+   * @return {string} Concatenated list of comments.
    */
-  this.ORDER_OVERRIDES = [];
+  allNestedComments(block) {
+    const comments = [];
+    const blocks = block.getDescendants(true);
+    for (let i = 0; i < blocks.length; i++) {
+      const comment = blocks[i].getCommentText();
+      if (comment) {
+        comments.push(comment);
+      }
+    }
+    // Append an empty string to create a trailing line break when joined.
+    if (comments.length) {
+      comments.push('');
+    }
+    return comments.join('\n');
+  }
 
   /**
-   * Whether the init method has been called.
-   * Generators that set this flag to false after creation and true in init
-   * will cause blockToCode to emit a warning if the generator has not been
-   * initialized. If this flag is untouched, it will have no effect.
-   * @type {?boolean}
+   * Generate code for the specified block (and attached blocks).
+   * The generator must be initialized before calling this function.
+   * @param {Block} block The block to generate code for.
```

Comment:
I know you didn't change this, but while you're here this needs a `?`. I guess this doesn't matter as long as the Typescript conversion tool that we eventually use still knows that this is nullable.

Reply:
Done.